File: amazon-clone/src/store/authStore.ts
import { create } from "zustand";
import { persist } from "zustand/middleware";
import { User, Order } from "@/lib/types";

interface AuthState {
  user: User | null;
  isAuthenticated: boolean;
  orders: Order[];
  signIn: (email: string, password: string) => boolean;
  signUp: (name: string, email: string, password: string) => boolean;
  signOut: () => void;
  addOrder: (order: Order) => void;
  cancelOrder: (orderId: string) => void;
}

export const useAuthStore = create<AuthState>()(
  persist(
    (set, get) => ({
      user: null,
      isAuthenticated: false,
      orders: [],

      signIn: (email: string, _password: string) => {
        // Mock authentication - accept any valid-looking email
        if (email && email.includes("@")) {
          const name = email.split("@")[0];
          set({
            user: {
              id: "user-1",
              name: name.charAt(0).toUpperCase() + name.slice(1),
              email,
            },
            isAuthenticated: true,
          });
          return true;
        }
        return false;
      },

      signUp: (name: string, email: string, _password: string) => {
        if (name && email && email.includes("@")) {
          set({
            user: {
              id: "user-" + Date.now(),
              name,
              email,
            },
            isAuthenticated: true,
          });
          return true;
        }
        return false;
      },

      signOut: () => {
        set({ user: null, isAuthenticated: false });
      },

      addOrder: (order: Order) => {
        set((state) => ({
          orders: [order, ...state.orders],
        }));
      },

      cancelOrder: (orderId: string) => {
        set((state) => ({
          orders: state.orders.map((order) =>
            order.id === orderId ? { ...order, status: "cancelled" as const } : order
          ),
        }));
      },
    }),
    {
      name: "novamart-auth",
    }
  )
);

File: ios-amazon-clone/NovaMart/Stores/AuthStore.swift
import Foundation
import Observation

@Observable
final class AuthStore {
    var user: User?
    var isAuthenticated: Bool = false
    var orders: [Order] = []

    init() {
        self.user = PersistenceService.load(User.self, forKey: PersistenceService.Keys.user)
        self.isAuthenticated = user != nil
        self.orders = PersistenceService.load([Order].self, forKey: PersistenceService.Keys.orders) ?? []
    }

    @discardableResult
    func signIn(email: String, password: String) -> Bool {
        guard !email.isEmpty, email.contains("@") else { return false }
        let name = String(email.split(separator: "@").first ?? "")
        let capitalizedName = name.prefix(1).uppercased() + name.dropFirst()
        let newUser = User(id: "user-1", name: capitalizedName, email: email)
        self.user = newUser
        self.isAuthenticated = true
        persist()
        return true
    }

    @discardableResult
    func signUp(name: String, email: String, password: String) -> Bool {
        guard !name.isEmpty, !email.isEmpty, email.contains("@") else { return false }
        let newUser = User(id: "user-\(Int(Date().timeIntervalSince1970))", name: name, email: email)
        self.user = newUser
        self.isAuthenticated = true
        persist()
        return true
    }

    func signOut() {
        self.user = nil
        self.isAuthenticated = false
        persist()
    }

    func addOrder(_ order: Order) {
        orders.insert(order, at: 0)
        persistOrders()
    }

    func cancelOrder(_ orderId: String) {
        guard let index = orders.firstIndex(where: { $0.id == orderId }),
              orders[index].status == .processing else { return }
        orders[index].status = .cancelled
        persistOrders()
    }

    private func persist() {
        if let user {
            PersistenceService.save(user, forKey: PersistenceService.Keys.user)
        } else {
            PersistenceService.remove(forKey: PersistenceService.Keys.user)
        }
    }

    private func persistOrders() {
        PersistenceService.save(orders, forKey: PersistenceService.Keys.orders)
    }
}
